File: packages/xyd-opencli2go/src/flags.ts
import { goBool, goSlice, goStr, goStruct, type GoVal } from './golit';
import type { FlagModel, GoType } from './model';

const FLAG_TYPE: Record<GoType, string> = {
  bool: 'BoolFlag',
  int: 'IntFlag',
  float: 'FloatFlag',
  slice: 'StringSliceFlag',
  string: 'StringFlag',
  json: 'StringFlag',
  file: 'StringFlag',
};

export function renderFlag(f: FlagModel): GoVal {
  const type = `cli.${FLAG_TYPE[f.goType]}`;
  const fields: [string, GoVal][] = [['Name', goStr(f.flagName)]];
  if (f.aliases.length) fields.push(['Aliases', goSlice('string', f.aliases.map(goStr))]);
  if (f.description) fields.push(['Usage', goStr(f.description)]);
  if (f.required) fields.push(['Required', goBool(true)]);
  if (f.hidden) fields.push(['Hidden', goBool(true)]);
  return goStruct(type, fields, true);
}

export function renderFlags(flags: FlagModel[]): GoVal[] {
  return flags.map(renderFlag);
}

File: crates/xyd_opencli2go/src/flags.rs
//! Flag rendering — port of flags.ts.

use crate::golit::{go_bool, go_slice, go_str, go_struct, GoVal};
use crate::model::FlagModel;

pub fn render_flag(f: &FlagModel) -> GoVal {
    let type_name = format!("cli.{}", f.go_type.flag_type());
    let mut fields: Vec<(String, GoVal)> = vec![("Name".to_string(), go_str(&f.flag_name))];
    if !f.aliases.is_empty() {
        let aliases = f.aliases.iter().map(|a| go_str(a)).collect();
        fields.push(("Aliases".to_string(), go_slice("string", aliases)));
    }
    if let Some(desc) = &f.description {
        fields.push(("Usage".to_string(), go_str(desc)));
    }
    if f.required {
        fields.push(("Required".to_string(), go_bool(true)));
    }
    if f.hidden {
        fields.push(("Hidden".to_string(), go_bool(true)));
    }
    go_struct(&type_name, fields, true)
}

pub fn render_flags(flags: &[FlagModel]) -> Vec<GoVal> {
    flags.iter().map(render_flag).collect()
}
